### PR TITLE
[nlminb] New package

### PR DESCRIPTION
**Pull request by kongdd**

### CODE REVIEW

File: N/nlminb/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+cd build
+cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_BUILD_TYPE=Release ..
+make -j${nproc}
+make install
+"""
+# -DBLAS_LIBRARIES="-l${LIBOPENBLAS}"
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = expand_cxxstring_abis(supported_platforms()) # build on all supported platforms
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:27]:
```suggestion
platforms = expand_gfortran_versions(supported_platforms()) # build on all supported platforms
```

Comment (kongdd) [reply N/nlminb/build_tarballs.jl:27]: Thank your timely suggestion. I will have a try. 
Another bug might at here https://github.com/geo-julia/nlminb.f/blob/master/CMakeLists.txt#L20. Could you give some suggestion about how to include libblas?

File: N/nlminb/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+if [[ "${nbits}" == 64 ]] && [[ "${target}" != aarch64* ]]; then
+    OPENBLAS="${libdir}/libopenblas64_.${dlext}"
+else
+    OPENBLAS="${libdir}/libopenblas.${dlext}"
+fi
+
+# -DBLAS_LIBRARIES=${libdir}/libopenblas.${dlext}
+CMAKE_FLAGS+=(-DCMAKE_INSTALL_PREFIX=${prefix}
+              -DBLAS_LIBRARIES=${OPENBLAS}
+              -DCMAKE_BUILD_TYPE=Release
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:27]: https://docs.binarybuilder.org/stable/build_tips/#CMake-builds
```suggestion
              -DCMAKE_BUILD_TYPE=Release
              -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}
```

**Comment by kongdd** [comment]
This package might be ready for merge. 
Thanks.

File: N/nlminb/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    FLAGS="-mfpmath=sse -msse2 -mstackrealign"
+fi
+CCFLAGS="-fPIC -DNDEBUG  -Iinclude -O2 -Wall -std=gnu99 ${FLAGS}"
+FCFLAGS="-fPIC -fno-optimize-sibling-calls -O2 ${FLAGS}"
+
+gfortran ${FCFLAGS} -c src/portsrc.f -o src/portsrc.o 
+gfortran ${FCFLAGS} -c src/d1mach.f -o src/d1mach.o 
+gcc ${CCFLAGS} -c src/port2.c -o src/port.o
+gfortran -shared -static-libgcc -lm -o libnlminb.${dlext} src/portsrc.o src/d1mach.o src/port.o lib/BLAS-3.8.0/blas_LINUX.a
+cp -L *.${dlext} ${libdir}
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:33]:
```suggestion
gfortran -shared -static-libgcc -lm -o ${libdir}/libnlminb.${dlext} src/portsrc.o src/d1mach.o src/port.o lib/BLAS-3.8.0/blas_LINUX.a
```

File: N/nlminb/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+"""
+# -DBLAS_LIBRARIES="-l${LIBOPENBLAS}"
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:35]:
```suggestion
```

File: N/nlminb/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+# platforms = expand_gfortran_versions(supported_platforms()) # build on all supported platforms
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:39]:
```suggestion
```

File: N/nlminb/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+else
+    FLAGS="-mfpmath=sse -msse2 -mstackrealign"
+fi
+CCFLAGS="-fPIC -DNDEBUG  -Iinclude -O2 -Wall -std=gnu99 ${FLAGS}"
+FCFLAGS="-fPIC -fno-optimize-sibling-calls -O2 ${FLAGS}"
+
+gfortran ${FCFLAGS} -c src/portsrc.f -o src/portsrc.o 
+gfortran ${FCFLAGS} -c src/d1mach.f -o src/d1mach.o 
+gcc ${CCFLAGS} -c src/port2.c -o src/port.o
+gfortran -shared -static-libgcc -lm -o libnlminb.${dlext} src/portsrc.o src/d1mach.o src/port.o lib/BLAS-3.8.0/blas_LINUX.a
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:32]: I'm a bit confused: what's this `lib/BLAS-3.8.0/blas_LINUX.a`?  A binary blob of a BLAS library, only for Linux and presumably only for a single platform?  I didn't follow all the attempts here, what's wrong with OpenBLAS or OpenBLAS32?

File: N/nlminb/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+]
+platforms = expand_gfortran_versions(platforms)
+platforms = platforms[2:end]
+
+## Linux x86_64 {libc=glibc, libgfortran_version=3.0.0}
+# This version not work: 
+# /blas_LINUX.a(dnrm2.o): relocation R_X86_64_PC32 against undefined symbol
+# `sqrt@@GLIBC_2.2.5' can not be used when making a shared object; recompile
+# with -fPIC/
+
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:55]:
```suggestion
```

File: N/nlminb/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libnlminb", :libnlminb),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    # Dependency("OpenBLAS_jll"),
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:63]:
```suggestion
```

File: N/nlminb/build_tarballs.jl (hunk 6 above)
Comment (kongdd) [reply N/nlminb/build_tarballs.jl:32]: Sorry. I have no idea how to include openblas correctly. So, build manually from tgz file. 
`blas` is downloaded at the official web, http://www.netlib.org/blas/blas-3.8.0.tgz. The `lib/BLAS-3.8.0/blas_LINUX.a` is the output of openblas library. 

Thanks for your comments and review.

**Comment by giordano** [comment]
I think the most straightforward thing is to build with `OpenBLAS32`, and this should work for all platforms without hassles:
```julia
using BinaryBuilder

name = "nlminb"
version = v"0.1.0"

# Collection of sources required to build NLopt
sources = [
    GitSource("https://github.com/geo-julia/nlminb.f.git",
              "6b2d460b866911e8bfca32b1217c08fb78c3111f"), # v0.1.0
]

# Bash recipe for building across all platforms
script = raw"""
cd $WORKSPACE/srcdir/nlminb.f

## compile by hand
if [[ "${proc_family}" == "intel" ]]; then
    FLAGS="-mfpmath=sse -msse2 -mstackrealign"
fi
CFLAGS="-fPIC -DNDEBUG  -Iinclude -O2 -Wall -std=gnu99 ${FLAGS}"
FFLAGS="-fPIC -fno-optimize-sibling-calls -O2 ${FLAGS}"

gfortran ${FFLAGS} -c src/portsrc.f -o src/portsrc.o 
gfortran ${FFLAGS} -c src/d1mach.f -o src/d1mach.o 
gcc ${CFLAGS} -c src/port2.c -o src/port.o
gfortran -shared -static-libgcc -lm -o ${libdir}/libnlminb.${dlext} src/portsrc.o src/d1mach.o src/port.o -lopenblas
"""

# These are the platforms we will build for by default, unless further
# platforms are passed in on the command line
platforms = expand_gfortran_versions(supported_platforms())

# The products that we will ensure are always built
products = [
    LibraryProduct("libnlminb", :libnlminb),
]

# Dependencies that must be installed before this package can be built
dependencies = [
    Dependency("CompilerSupportLibraries_jll"),
    Dependency("OpenBLAS32_jll"),
]

# Build the tarballs, and possibly a `build.jl` as well.
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```

File: N/nlminb/build_tarballs.jl [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    LibraryProduct("libnlminb", :libnlminb),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency("CompilerSupportLibraries_jll")
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```
Comment (giordano) [inline N/nlminb/build_tarballs.jl:58]: I think the most straightforward thing is to build with `OpenBLAS32`, and this should work for all platforms without hassles:
```suggestion
using BinaryBuilder

name = "nlminb"
version = v"0.1.0"

# Collection of sources required to build NLopt
sources = [
    GitSource("https://github.com/geo-julia/nlminb.f.git",
              "6b2d460b866911e8bfca32b1217c08fb78c3111f"), # v0.1.0
]

# Bash recipe for building across all platforms
script = raw"""
cd $WORKSPACE/srcdir/nlminb.f

## compile by hand
if [[ "${proc_family}" == "intel" ]]; then
    FLAGS="-mfpmath=sse -msse2 -mstackrealign"
fi
CFLAGS="-fPIC -DNDEBUG  -Iinclude -O2 -Wall -std=gnu99 ${FLAGS}"
FFLAGS="-fPIC -fno-optimize-sibling-calls -O2 ${FLAGS}"

gfortran ${FFLAGS} -c src/portsrc.f -o src/portsrc.o 
gfortran ${FFLAGS} -c src/d1mach.f -o src/d1mach.o 
gcc ${CFLAGS} -c src/port2.c -o src/port.o
gfortran -shared -static-libgcc -lm -o ${libdir}/libnlminb.${dlext} src/portsrc.o src/d1mach.o src/port.o -lopenblas
"""

# These are the platforms we will build for by default, unless further
# platforms are passed in on the command line
platforms = expand_gfortran_versions(supported_platforms())

# The products that we will ensure are always built
products = [
    LibraryProduct("libnlminb", :libnlminb),
]

# Dependencies that must be installed before this package can be built
dependencies = [
    Dependency("CompilerSupportLibraries_jll"),
    Dependency("OpenBLAS32_jll"),
]

# Build the tarballs, and possibly a `build.jl` as well.
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```

**Comment by giordano** [comment]
Ok, I think we're good to go now!

**Comment by kongdd** [comment]
@giordano Thanks for your revision.

**Pull request merged**